### ISSUE TITLE
[SCCP] Missed optimization : better range for overflow operators with `nuw`/`nsw`

### ISSUE THREAD

**Issue by XChy** [llvm/llvm-project#80159]
Alive2 proof: https://alive2.llvm.org/ce/z/6AfwYo

### Motivating example 

```llvm
define i1 @src(i32 %a) {
entry:
  %cmp = icmp ne i32 %a, 0
  br i1 %cmp, label %then, label %else
then:
  %mul = mul nuw nsw i32 %a, 10000 ; Expected: [10000, UINT_MAX], Actual: Overdefined
  %add = add nsw i32 %mul, -5000   ; Expected: [5000, UINT_MAX], Actual: Overdefined
  %cond = icmp ult i32 %add, 4999
  ret i1 %cond
else:
  call void @dummy()
  ret i1 0
}
```
`%cond`  can be optimized away:
```llvm
define i1 @tgt(i32 %a) {
entry:
  %cmp = icmp ne i32 %a, 0
  br i1 %cmp, label %then, label %else
then:
  ret i1 0
else:
  call void @dummy()
  ret i1 0
}
```

Not looking into SCCP yet, but I think such missed optimization is the same for other overflow operators like `sub` and `shl`.
NOTE: The overflow operators in original real-world ir can only be optimized at use, just like [this one](https://alive2.llvm.org/ce/z/g6_6fr). CVP/LVI also missed it, but I think it's easier to solve SCCP's one firstly, so post this issue.

### Real-world motivation

This snippet of IR is derived from [jemalloc/src/fxp.c@fxp_print](https://github.com/jemalloc/jemalloc/blob/f96010b7fa8ce5f83802144bdebf2bb7a6679649/src/fxp.c#L98) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/hE3vza9Kc

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

**Comment by fhahn** [comment]
>   %mul = mul nuw nsw i32 %a, 10000 ; Expected: [10000, UINT_MAX], Actual: Overdefined

For this particular case yes, but that depends on the uses of `%mul`. It could also be poison. But ValueLatticeElement can represent ranges that may contain undef/poison

**Comment by XChy** [comment]
> > %mul = mul nuw nsw i32 %a, 10000 ; Expected: [10000, UINT_MAX], Actual: Overdefined
> 
> For this particular case yes, but that depends on the uses of `%mul`. It could also be poison. But ValueLatticeElement can represent ranges that may contain undef/poison

If I don't miss something, SCCP doesn't preserve undef/poison when both LHS and RHS are constant ranges currently. Instead, it seems to use `ValueLatticeElement::getRange` directly, doing refinement now. I'm not sure what precision problem this brings, but it looks at least correct to me.
https://github.com/llvm/llvm-project/blob/9acd61ec1999ac3a54371d0a8b9d922ef5ca2b50/llvm/lib/Transforms/Utils/SCCPSolver.cpp#L1490

**Comment by nikic** [comment]
Yes, ValueLattice is always "correct up to poison" only. Only undef requires special handling.

**Comment by nikic** [comment]
@antoniofrighetto Are you still working on this?

**Comment by antoniofrighetto** [comment]
@nikic, I have a patch attempting to add `multiplyWithNoWrap`, but I'm still getting some failures when exhaustively testing (specifically when handling overflown cases in `NoUnsignedWrap`, and testing it via `umul_ov`). Haven't managed to look into, but I'd like to investigate better this weekend, if possible.

**Comment by nikic** [comment]
I'd suggest implementing the SCCP part of this issue first.

**Comment by antoniofrighetto** [comment]
SCCP leverages the results of ConstantRange though. IIUC SCCP should just be extended from:

https://github.com/llvm/llvm-project/blob/9acd61ec1999ac3a54371d0a8b9d922ef5ca2b50/llvm/lib/Transforms/Utils/SCCPSolver.cpp#L1489-L1490

To something like:
```c++
  auto *BO = cast<BinaryOperator>(&I);
  ConstantRange R = ConstantRange::getEmpty(I.getType()->getScalarSizeInBits());
  if (auto *OBO = dyn_cast<OverflowingBinaryOperator>(BO)) {
    bool HasNUW = OBO->hasNoUnsignedWrap();
    bool HasNSW = OBO->hasNoSignedWrap();
    unsigned Flags = (HasNUW ? OverflowingBinaryOperator::NoUnsignedWrap : 0) |
                     (HasNSW ? OverflowingBinaryOperator::NoSignedWrap : 0);
    R = A.overflowingBinaryOp(BO->getOpcode(), B, Flags);
  } else {
    R = A.binaryOp(BO->getOpcode(), B);
  }
  mergeInValue(&I, ValueLatticeElement::getRange(R));
  ```
  And have the logic appropriately handled in ConstantRange.

**Comment by nikic** [comment]
@antoniofrighetto Yes. This will already handle nowrap add/sub. Adding support for nowrap mul is an independent change on top of that.
